data_steward/tools/purge_hpo_data.py

```diff
@@ -0,0 +1,100 @@
+"""
+Empty all CDM tables associated with one or more HPO sites from an EHR dataset
+
+This can be used to exclude sites' data from the EHR dataset before 
+generating the unioned dataset which will be used for the CDR.
+
+NOTE: This does not affect non-clinical tables (e.g. achilles_*)
+"""
+import logging
+from typing import List
+
+from google.cloud import bigquery
+
+from utils import bq
+import bq_utils
+import resources
+
+LOGGER = logging.getLogger(__name__)
+LIST_TABLES_MAX_RESULTS = 5000
+DELETE_QUERY_TPL = bq.JINJA_ENV.from_string("""
+{%- for table in tables_to_purge -%}
+DELETE FROM `{{table.project}}.{{table.dataset_id}}.{{table.table_id}}` WHERE 1=1;
+{%- endfor -%}
+""")
+
+
+def _filter_hpo_tables(tables: List[bigquery.table.TableListItem],
+                       hpo_id: str) -> List[bigquery.table.TableListItem]:
+    """
+    Given a list of tables get those associated with an HPO submission
+
+    :param tables: list of tables to filter
+    :param hpo_id: identifies the HPO
+    :return: list of tables assoicated with the HPO
+    """
+    expected_tables = [
+        bq_utils.get_table_id(hpo_id, table) for table in resources.CDM_TABLES
+    ]
+    return [table for table in tables if table.table_id in expected_tables]
+
+
+def purge_hpo_data(client: bigquery.Client, dataset: bigquery.DatasetReference,
+                   hpo_ids: List[str]) -> bigquery.QueryJob:
+    """
+    Empty all CDM tables associated with one or more HPO sites
+    
+    :param client: Active bigquery client object 
+    :param dataset: the dataset to purge data from
+    :param hpo_ids: Identifies the HPO sites whose data should be purged
+    :return: Query job associated with removing all the records
+    :raises RuntimeError if CDM tables associated with a site are not found in the dataset
+    """
+    LOGGER.debug(
+        f'purge_hpo_data called with dataset={dataset.dataset_id} and hpo_ids={hpo_ids}'
+    )
+    all_tables = list(
+        client.list_tables(dataset=dataset,
+                           max_results=LIST_TABLES_MAX_RESULTS))
```

Comment:
I can no longer find the doc page that said the default `max_results`- but can you believe it's a measly 50! Anyway I will follow your first suggestion and leave a comment in there for now.

I'd also like to avoid a hard coded value, but I'm conflicted about how to go about it. The caller's intent is always to purge HPO data from "all tables where they are found." A user-supplied `max_results` parameter would be far removed from the intent and it can be difficult to predict what would happen when the value provided is less than the number of tables in the dataset.

Here are some other alternatives I'd like to discuss with you (**after** this CDR gen).
1. Determine `max_results` based on metadata query
```python
  # This query should be free of cost as compared to INFORMATION_SCHEMA.TABLES
  client.query(f"SELECT COUNT(1) max_results FROM `{project_id}.{dataset_id}.__TABLES__`")
```
2. Forgo `max_results` altogether and instead generate and execute dynamic query in a script.
A. Get the tables based on `hpo_id` prefix.
    ```sql
    DECLARE qs DEFAULT(
    ARRAY(
        -- assuming my_dataset and my_site are filled e.g. using jinja
        SELECT FORMAT('DELETE FROM %s.%s.%s', table_catalog, table_schema, table_name) 
        FROM `my_dataset.INFORMATION_SCHEMA.COLUMNS`
        WHERE column_name='person_id' AND table_name LIKE 'my_site_%'
    )
    );
    DECLARE n DEFAULT(ARRAY_LENGTH(qs));
    DECLARE i DEFAULT(0);
    DECLARE q STRING;
    WHILE i < n DO
    SET q = qs[ARRAY_OFFSET(i)];
    EXECUTE IMMEDIATE q;
    SET i = i + 1;
    END WHILE;
    ```
    B. Same idea as A but get the tables based on labels we apply e.g. `{"hpo_id": "some_hpo"}` and avoid the weirdness when one `hpo_id` is a substring of another (e.g. `site_a` &#8800; `site_a_1` &#8800; `site_a_1_b`).

Reply:
I think we both agree option 1 is better than a hard coded value and possibly more maintainable than option 2.